Add support for additional APIs, test coverage

kujenga opened the pull request:
This PR adds support for a few more APIs from https://docs.helium.com/api and adds test coverage for them (and others) too. I'm new-ish to Rust to happy to iterate on any feedback!

madninja commented on src/lib.rs:
```diff
@@ -133,9 +164,13 @@ impl Client {
+        return Ok(self.fetch_data(request_url)?.data);
+    }
+
+    pub(crate) fn fetch_data<T: DeserializeOwned>(&self, url: String) -> error::Result<Data<T>> {
```
Make url a `&str`?

madninja commented on src/lib.rs:
```diff
@@ -159,6 +194,11 @@ impl Client {
     }
 
+    /// Get wallet information for the richest accounts
+    pub fn get_accounts_richest(&self, limit: Option<u64>) -> error::Result<Vec<Account>> {
```
a u64 for limit is perhaps a bit overlay optimistic at how large pages can be.. u16 or u32 at best

madninja commented on src/lib.rs:
```diff
@@ -175,6 +215,56 @@ impl Client {
+    ) -> error::Result<(Vec<Reward>, Option<String>)> {
+        let base = Url::parse(&format!("{}/", self.base_url)).unwrap();
+        let mut request_url = base.join(&format!("hotspots/{}/rewards", address)).unwrap();
+        for (k, v) in [
```
Try using `Url::parse_with_params` after creating a vec of options that have values.

kujenga replied:
Nice, I've added a `build_url` private helper to encapsulate that and the "optional param" logic, swapping out the previous implementation.

madninja commented on src/lib.rs:
```diff
@@ -175,6 +215,56 @@ impl Client {
+        .iter()
+        {
+            if v.is_some() {
+                request_url.query_pairs_mut().append_pair(k, v.unwrap());
```
See above.. I think parse_with_params may be a cleaner solution. If not, at least hang on to the results of query_pairs_mut so you can call append_pair without re-requesting the mutable query pairs every time?

kujenga replied:
This now uses the `build_url` helper

madninja commented on src/lib.rs:
```diff
@@ -175,6 +215,56 @@ impl Client {
+        Ok((data.data, data.cursor))
+    }
+
+    pub fn get_oracle_price_current(&self) -> error::Result<OraclePrice> {
```
match API name -> `get_oracle_prices_current`

madninja reviewed (changes requested):
Thanks for your PR!! 

A few changes requested but this is looking pretty good!

kujenga commented:
I believe I've addressed all the feedback, thanks! Not sure what's happening with the build, I'm not able to see what the failures are (there's no dropdown for output), and I'm developing on mac and see no issues locally. Maybe it just needs to be re-run?

<img width="749" alt="Screen Shot 2021-02-23 at 19 11 37" src="https://user-images.githubusercontent.com/4854867/108926572-b0a1a200-760c-11eb-9760-a1a9786b34c1.png">

madninja commented on src/lib.rs:
```diff
@@ -87,6 +88,35 @@ pub struct Hotspot {
+
+impl OraclePrice {
+    /// The price submitted by the oracle in USD.
+    pub fn price_usd(&self) -> f64 {
```
rename to `to_usd()`

madninja commented on src/lib.rs:
```diff
@@ -131,11 +162,27 @@ impl Client {
+    pub(crate) fn build_url(&self, path: &str, params: &[(&str, Option<&str>)]) -> reqwest::Url {
+        // Create a filtered set of string params that were specified.
+        let params = params.iter().filter_map(|p| {
+            if p.1.is_none() {
```
this is neat.. There has to be a more idiomatic way to do this though?

kujenga replied:
Sure, happy to iterate here, though some guidance on what direction is most idiomatic would be most welcome! Two possibilities come to mind:

Clarify flow a bit by splitting out the filter/map stages:

```rust
        let params = params
            .iter()
            .filter(|p| p.1.is_some())
            .map(|p| (p.0, p.1.unwrap()));
```

Or this is how I would write it if I was trying to write idiomatic Go code:

```rust
        let mut actual_params = Vec::new();
        for p in params {
            if p.1.is_some() {
                actual_params.push((p.0, p.1.unwrap()));
            }
        }
```

If something else entirely comes to mind let me know.

kujenga replied:
I've pushed the first one as a starting point

kujenga commented:
Rebased, and attempted to address the comments 👍

madninja commented:
I figured it may be easier to show what I'm working on: https://github.com/helium/helium-api-rs/pull/24 and see how we can get your additions into this newer proposed approach?

kujenga commented:
Ah, neat! Yeah happy to re-work this on top of those changes. Once that's in I can take another pass here.

madninja commented:
I'm going to close this since the new main branch I believe has what was in this branch. Thanks for triggering me to finish it of @kujenga !
